core/toolbox.js

```diff
@@ -252,6 +252,163 @@ Blockly.Toolbox.prototype.renderTree = function(languageTree) {
   }
 };
 
+/**
+ * Create the toolbox tree.
+ * @param {Array.<Blockly.utils.toolbox.ToolboxInfo>} toolboxDef List of objects
+ *    holding information on toolbox contents.
+ * @param {!Blockly.tree.BaseNode} treeOut The TreeControl or TreeNode
+ *     object built from the childNodes.
+ * @return {Blockly.tree.BaseNode} TreeNode to open at startup (or null).
+ * @private
+ */
+Blockly.Toolbox.prototype.createTree_ = function(toolboxDef, treeOut) {
+  var openNode = null;
+  var lastElement = null;
+  if (!toolboxDef) {
+    return null;
+  }
+
+  for (var i = 0, childIn; (childIn = toolboxDef[i]); i++) {
+    switch (childIn.kind.toUpperCase()) {
+      case 'CATEGORY':
+        var categoryInfo = /** @type {Blockly.utils.toolbox.CategoryInfo} */ (childIn);
+        openNode = this.addCategory_(categoryInfo, treeOut) || openNode;
+        lastElement = childIn;
+        break;
+      case 'SEP':
+        var separatorInfo = /** @type {Blockly.utils.toolbox.SeparatorInfo} */ (childIn);
+        lastElement = this.addSeparator_(separatorInfo, treeOut, lastElement) || lastElement;
+        break;
+      case 'BLOCK':
+      case 'SHADOW':
+      case 'LABEL':
+      case 'BUTTON':
+        treeOut.contents.push(childIn);
+        lastElement = childIn;
+        break;
+    }
+  }
+  return openNode;
+};
+
+/**
+ * Add a category to the toolbox tree.
+ * @param {!Blockly.utils.toolbox.CategoryInfo} categoryInfo The object holding
+ *    information on the category.
+ * @param {!Blockly.tree.BaseNode} treeOut The TreeControl or TreeNode
+ *     object built from the childNodes.
+ * @return {Blockly.tree.BaseNode} TreeNode to open at startup (or null).
+ * @private
+ */
+Blockly.Toolbox.prototype.addCategory_ = function(categoryInfo, treeOut) {
+  var openNode = null;
+  // Decode the category name for any potential message references
+  // (eg. `%{BKY_CATEGORY_NAME_LOGIC}`).
+  var categoryName = Blockly.utils.replaceMessageReferences(categoryInfo['name']);
+  var childOut = this.addCategoryChild_(categoryName, treeOut);
+  var custom = categoryInfo['custom'];
+
+  if (custom) {
+    // Variables and procedures are special dynamic categories.
+    childOut.contents = custom;
+  } else {
+    openNode = this.createTree_(categoryInfo.contents, childOut) || openNode;
+  }
+  this.addColourOrStyle_(categoryInfo, childOut, categoryName);
+  openNode = this.setExpanded_(categoryInfo, childOut) || openNode;
+  return openNode;
+};
+
+/**
+ * Create the TreeNode for the category.
```

Comment:
What's the difference between `addCategory` and `addCategoryChild`?

Reply:
removed addCategoryChild. And just put the logic in addCategory.